tsdb/series_file.go

```diff
@@ -690,196 +530,202 @@ func (a seriesKeys) Less(i, j int) bool {
-	binary.BigEndian.PutUint64(buf[0:8], uint64(n))
-	binary.BigEndian.PutUint64(buf[8:16], uint64(len(src)))
+		// If the existing elem has probed less than us, then swap places with
+		// existing elem, and keep going to find another slot for that elem.
+		if d := rhh.Dist(elemHash, pos, capacity); d < dist {
 
-	// Loop over all series in data. Offset starts at 1.
-	for b, offset := src[1:], uint64(1); len(b) > 0; {
-		var key []byte
-		key, b = ReadSeriesKey(b)
+			// Insert current values.
+			binary.BigEndian.PutUint64(elem[:8], uint64(offset))
+			binary.BigEndian.PutUint64(elem[8:], id)
 
-		insertSeriesMap(src, buf, key, offset, capacity)
-		offset += uint64(len(key))
-	}
+			// Swap with values in that position.
+			hash, key, offset, id = elemHash, elemKey, elemOffset, elemID
 
-	return buf
+			// Update current distance.
+			dist = d
+		}
+	}
 }
 
-func insertSeriesMap(src, buf, key []byte, val uint64, capacity int64) {
-	mask := int64(capacity - 1)
-	hash := rhh.HashKey(key)
+func (c *SeriesFileCompactor) insertIDOffsetMap(dst []byte, capacity int64, id uint64, offset int64) {
+	mask := capacity - 1
+	hash := rhh.HashUint64(id)
 
 	// Continue searching until we find an empty slot or lower probe distance.
-	for dist, pos := int64(0), hash&mask; ; dist, pos = dist+1, (pos+1)&mask {
-		elem := buf[SeriesMapHeaderSize+(pos*SeriesMapElemSize):]
-		elem = elem[:SeriesMapElemSize]
+	for i, dist, pos := int64(0), int64(0), hash&mask; ; i, dist, pos = i+1, dist+1, (pos+1)&mask {
+		elem := dst[(pos * SeriesIndexElemSize):]
 
-		h := int64(binary.BigEndian.Uint64(elem[:8]))
-		v := binary.BigEndian.Uint64(elem[8:])
-		k, _ := ReadSeriesKey(src[v:])
-
-		// Empty slot found or matching key, insert and exit.
-		if h == 0 || bytes.Equal(key, k) {
-			binary.BigEndian.PutUint64(elem[:8], uint64(hash))
-			binary.BigEndian.PutUint64(elem[8:], val)
+		// If empty slot found or matching id, insert and exit.
+		elemID := binary.BigEndian.Uint64(elem[:8])
+		elemOffset := int64(binary.BigEndian.Uint64(elem[8:]))
+		if elemOffset == 0 || elemOffset == offset {
+			binary.BigEndian.PutUint64(elem[:8], id)
+			binary.BigEndian.PutUint64(elem[8:], uint64(offset))
 			return
 		}
 
+		// Hash key.
+		elemHash := rhh.HashUint64(elemID)
+
 		// If the existing elem has probed less than us, then swap places with
 		// existing elem, and keep going to find another slot for that elem.
-		if d := rhh.Dist(h, pos, capacity); d < dist {
+		if d := rhh.Dist(elemHash, pos, capacity); d < dist {
 			// Insert current values.
-			binary.BigEndian.PutUint64(elem[:8], uint64(hash))
-			binary.BigEndian.PutUint64(elem[8:], val)
+			binary.BigEndian.PutUint64(elem[:8], id)
+			binary.BigEndian.PutUint64(elem[8:], uint64(offset))
 
 			// Swap with values in that position.
-			hash, key, val = h, k, v
+			hash, id, offset = elemHash, elemID, elemOffset
 
 			// Update current distance.
 			dist = d
 		}
+
+		if i > capacity {
+			panic("rhh map full")
+		}
```

Comment:
The `hashReader` function further down in this file can be deleted.

Reply:
Fixed in 3900c94.